numpy/core/_exceptions.py

```diff
@@ -122,21 +122,105 @@ class TooHardError(RuntimeError):
 
 @set_module('numpy')
 class AxisError(ValueError, IndexError):
-    """ Axis supplied was invalid. """
+    """Axis supplied was invalid.
+
+    Raised whenever an ``axis`` parameter is specified that is larger than
+    the number of array dimensions.
+
+    .. versionadded:: 1.22
+
+        Added the ``axis`` and ``ndim`` attributes.
+
+    Parameters
+    ----------
+    axis : int or str
+        The out of bounds axis or a custom exception message.
+        If an axis is provided, then `ndim` should be specified as well.
+    ndim : int, optional
+        The number of array dimensions.
+    msg_prefix : str, optional
+        A prefix for the exception message.
+
+    Attributes
+    ----------
+    axis : int, optional
+        The out of bounds axis or ``None`` if a custom exception
+        message was provided. This should be the axis as passed by
+        the user, before any normalization to resolve negative indices.
+    ndim : int, optional
+        The number of array dimensions or ``None`` if a custom exception
+        message was provided.
+
+    Examples
+    --------
+    >>> array_1d = np.arange(10)
+    >>> np.cumsum(array_1d, axis=1)
+    Traceback (most recent call last):
+      ...
+    numpy.AxisError: axis 1 is out of bounds for array of dimension 1
+
+    Negative axes are preserved:
+
+    >>> np.cumsum(array_1d, axis=-2)
+    Traceback (most recent call last):
+      ...
+    numpy.AxisError: axis -2 is out of bounds for array of dimension 1
+
+    The class constructor generally takes the axis and arrays'
+    dimensionality as arguments:
+
+    >>> np.AxisError(2, 1, msg_prefix='error')
+    AxisError('error: axis 2 is out of bounds for array of dimension 1')
+
+    Alternatively, a custom exception message can be passed:
+
+    >>> np.AxisError('Custom error message')
+    AxisError('Custom error message')
+
+    """
+
+    __slots__ = ("axis", "ndim")
+
     def __init__(self, axis, ndim=None, msg_prefix=None):
         # single-argument form just delegates to base class
         if ndim is None and msg_prefix is None:
             msg = axis
+            self.axis = None
+            self.ndim = None
 
         # do the string formatting here, to save work in the C code
         else:
             msg = ("axis {} is out of bounds for array of dimension {}"
                    .format(axis, ndim))
             if msg_prefix is not None:
                 msg = "{}: {}".format(msg_prefix, msg)
```

Comment:
Another option would be to follow the pattern of https://github.com/python/cpython/blob/79db11ce99332d62917be9d03b31494b1ff2f96a/Lib/subprocess.py#L60, and not compute the message at all here and do it inside `__str__` instead; that might make pickling more straightforward.

Are there any examples of the `__reduce__` pattern you use in Exceptions in the standard library?

Reply:
> Another option would be to follow the pattern 

I'm not sure if this pattern would alleviate the issue. Namelly, we'd still have to parse the arguments in order to assign a sensible value to the `axis` and `ndim` attributes, making it likelly that we're still stuck with more or less the same pattern inside `__init__`.

> Are there any examples of the `__reduce__` pattern you use in Exceptions in the standard library?

There are a couple of them (see [`multiprocessing.pool.ExceptionWithTraceback`](https://github.com/python/cpython/blob/bb3e0c240bc60fe08d332ff5955d54197f79751c/Lib/multiprocessing/pool.py#L63-L70) for example). 

Most of the time custom `__reduce__` implementations follow a `(cls, cls_args_tuple)` pattern, but occasionailly a dedicated reconstructor function is used rather than the type object itself. More specifically, whenever there is some sort of signature mismatch between the class constructor and the to-be pickled attributes.

Reply:
This gives us a nice mechanism to search for them:

```python
def all_subclasses(e):
    yield e
    for s in e.__subclasses__():
        yield from all_subclasses(s)

custom_str = [e for e in all_subclasses(Exception) if e.__str__ != Exception.__str__]
custom_reduce = [e for e in all_subclasses(Exception) if e.__reduce__ != Exception.__reduce__ ]
```